compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -2080,7 +2080,18 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                 )
             }
             hir::TyKind::TraitObject(bounds, lifetime, repr) => {
-                self.prohibit_or_lint_bare_trait_object_ty(hir_ty, in_path);
+                if let Some(guard) = self.prohibit_or_lint_bare_trait_object_ty(hir_ty, in_path) {
+                    if let hir::Node::Item(_) | hir::Node::TraitItem(_) | hir::Node::ImplItem(_) =
+                        tcx.parent_hir_node(hir_ty.hir_id)
+                    {
+                        // `tests/ui/suggestions/issue-116434-2021.rs` and others would otherwise
+                        // fail with no error beign emitted. We restrict the silencing only to
```

Comment:
What do you mean by "fail with no error [being] emitted"? In this branch we have `guard : ErrorGuaranteed`, we should've emitted at least one error?

Reply:
For some reason (which I don't understand, as it goes against my understanding of how this is meant to work), the stashed error ends up never getting emitted, which of course causes an ICE for an error being constructed but not emitted. This check does side step that.

Reply:
nnethercote has dug into this a while back and written a nice explainer: https://github.com/rust-lang/rust/issues/120482#issuecomment-1937991551